Action Bar: add following, liking, and more options modal

creativecoder opened the pull request:
#### Changes proposed in this Pull Request:

<img width="459" alt="image" src="https://user-images.githubusercontent.com/1699996/192343643-e614b8d1-3320-4c4f-8ba0-24913c21cb10.png">


- Adds a working like button for single posts
- Adds a working follow button for the site
- Adds modal with more actions

Combine with D87870-code to test liking and following

Fixes #27356 (dropping a CSS var not available in 2023 theme)

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

pe7F0s-1U-p2

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

Apply D87870-code and sandbox the specified domains

- Switch to this branch and build the project: jetpack install --all
- Build the action bar package: jetpack build packages/action-bar
- Activate the Action bar module from /wp-admin/admin.php?page=jetpack_modules
- Open a single post page on the front-end, and you should see the action bar in the lower left corner
- Click on the dots to see more options
- Click on the star to like or unlike the post
- Click on the plus to follow or unfollow the site

You can check likes by visiting https://wordpress.com/activities/likes

You can check follows by visiting https://wordpress.com/following/manage

Additional testing

- With a wpcom sandbox, modify the api to return an error response.
- You should see an error message, like the one below

<img width="487" alt="image" src="https://user-images.githubusercontent.com/1699996/192343975-7f14465e-8c71-4a1f-8cb8-b9e5960fbe29.png">

- Test with third-party cookies disabled (e.g. use a Chrome incognito window)... liking and following should still work

Note there is a known issue if testing with *.jurassic.ninja or *.jurassic.tube that follows the parent site instead of the subdomain, see D88262-code.

creativecoder committed:
[not verified] Try an action bar with iframed like button

roo2 committed:
[not verified] Add more options modal

roo2 committed:
[not verified] changelog

roo2 commented:
I added the UI and modal code back into jetpack, it was pretty smooth sailing, hopefully we've found the right home for everything.

creativecoder committed:
[not verified] A bit of cleanup

- Use correct escaping functions for urls, attrs, html
- Get the mapped domain and url for Simple sites

creativecoder committed:
[not verified] Display blog name, or fallback to domain

creativecoder committed:
[not verified] Abstract is_active method

creativecoder commented:
- Added a working like button (when combined with wpcom diff)
- Added testing instructions to the PR description
- Merged that latest from trunk

creativecoder commented on projects/packages/action-bar/src/js/utils.js:
```diff
@@ -0,0 +1,126 @@
+			el.classList.add( className );
+		} else {
+			el.className += ' ' + className;
+		}
+	}
+}
+
+/**
+ * Check if class exists
+ *
+ * @param {Element} el - element
+ * @param {string} className - class to check for
+ * @returns {boolean} true if class exists
+ */
+export function _classHas( el, className ) {
+	if ( ! el || ! className ) {
+		return false;
+	}
+	return el.classList.contains( className );
+}
+
+/**
+ * Remove class from element
+ *
+ * @param {Element} el - element
+ * @param {string} className - class to remove
+ * @returns {boolean} status
+ */
+export function _classRemove( el, className ) {
+	if ( ! _exists( el ) || ! className ) {
+		return false;
+	}
+	if ( el.classList ) {
+		return el.classList.remove( className );
+	}
+	return ( el.className = el.className.replace(
+		new RegExp( '(^|\\b)' + className.split( ' ' ).join( '|' ) + '(\\b|$)', 'gi' ),
+		' '
+	) );
+}
```
I don't think we need these for manipulating classes. classList.add, remove, and contains seems to have good enough browser support now.

roo2 replied:
I simplified that 👍

roo2 committed:
[not verified] address PR feedback, fix report URL, add fallback if favicon doesn't exist

roo2 commented:
I added a fallback for the favicon url, you can test by changing the favicon url to something that won't work like `favicon12345.ico`. It was finicky to get it working, I based it on this approach https://stackoverflow.com/questions/980855/inputting-a-default-image-in-case-the-src-attribute-of-an-html-img-is-not-vali

creativecoder committed:
[not verified] Adds follow button to action bar

roo2 committed:
[not verified] convert to scss

roo2 committed:
[not verified] update changelog and package version

roo2 committed:
[not verified] handle errors

creativecoder committed:
Change position to use px, in case wp spacing variable is not set

roo2 committed:
tweak style of error notification

creativecoder committed:
[not verified] Specify top level Feedbag class for wpcom compatibility

creativecoder committed:
Update enqueue priority to allow overriding existing wpcom actionbar

roo2 committed:
fix modal shade on mobile

roo2 commented on projects/packages/action-bar/src/class-action-bar.php:
```diff
@@ -46,18 +47,111 @@ public function enqueue_scripts() {
+							<?php comment_icon( __( 'Leave a comment', 'jetpack-action-bar' ) ); ?>
+						</button>
+					</li>
+					<!-- TODO: remove iframe focus state and tab directly to buttons in iframe, if possible -->
```
is this todo done now? tabbing seems to work as expected

roo2 commented on projects/packages/action-bar/src/class-action-bar.php:
```diff
@@ -46,18 +47,111 @@ public function enqueue_scripts() {
+			<div class="jetpack-action-bar__modal">
+				<header>
+					<object data="/favicon.ico" class="site-icon" type="image/x-icon" >
+						<?php globe_icon( __( 'site icon', 'jetpack-action-bar' ) ); ?>
```
I tested this fallback by changing the name `/favicon.ico` to something that doesn't exist
<img width="378" alt="Screen Shot 2022-10-04 at 5 16 24 pm" src="https://user-images.githubusercontent.com/22446385/193757941-520ad724-f289-4213-8751-afcca8a441ec.png">

roo2 commented on projects/packages/action-bar/src/class-action-bar.php:
```diff
@@ -46,18 +47,111 @@ public function enqueue_scripts() {
+					<a href="#" class="jetpack-action-bar__close close"><?php close_icon( __( 'close', 'jetpack-action-bar' ) ); ?></a>
+				</header>
+				<section class="menu">
+					<a href="https://wordpress.com/abuse/?report_url=<?php echo rawurlencode( $post_url ); ?>"><?php esc_html_e( 'Report this content', 'jetpack-action-bar' ); ?></a>
```
hmm I wonder if it makes sense to "report" non wpcom blogs?

roo2 committed:
tidy up js and php

roo2 commented on projects/packages/action-bar/src/scss/action-bar.scss:
```diff
@@ -17,6 +37,602 @@
+	padding: 8px;
+	margin: 0;
+}
+/* .action-bar {
```
I left in some commented out css because we are basing our CSS on work @davemart-in did in an HMTL prototype 2dcb4-pb, it will be enabled as we add the features it corresponds to.

roo2 reviewed (commented):
LGTM (if I do say so myself) let's get some feedback from jetpack crew

anomiex committed:
Merge branch 'trunk' into add/action-bar-iframed-like-button

anomiex commented on projects/packages/action-bar/src/action-bar-icons.php:
```diff
@@ -9,15 +22,22 @@
 		xmlns="http://www.w3.org/2000/svg"
 	>
-		<title>${ jetpackActionBar.commentTitle }</title>
+		<title><?php echo esc_html( $title ); ?></title>
```
I wonder if it would be safer to use `esc_xml` instead of `esc_html`, to avoid the possibility that some filter introduces HTML entities that are not valid in SVG.

Same for the other functions below.

anomiex commented on projects/packages/action-bar/src/action-bar-icons.php:
```diff
@@ -68,23 +95,38 @@
+</svg>';
+	?>
+		<div class="jetpack-action-bar__globe-icon">
+			<img srcset="data:image/svg+xml,<?php echo rawurlencode( $globe_html ); ?>" alt="<?php echo esc_html( $title ); ?>"/>
```
```suggestion
			<img srcset="data:image/svg+xml,<?php echo rawurlencode( $globe_html ); ?>" alt="<?php echo esc_attr( $title ); ?>"/>
```

anomiex commented on projects/packages/action-bar/src/class-action-bar.php:
```diff
@@ -46,18 +47,111 @@ public function enqueue_scripts() {
+		$blog_name  = get_bloginfo( 'name' );
+		$post_url   = get_permalink( $post_id );
+		$reader_url = $this->get_reader_url( $blog_id );
+		$widget_src = sprintf( 'https://widgets.wp.com/action-bar/#blog_id=%2$d&amp;post_id=%3$d&amp;origin=%1$s://%4$s', $protocol, $blog_id, $post_id, $domain );
```
The domain should be urlencoded, just in case there's something odd going on.

As for the ampersands, it's probably better to let `esc_url()` do it later than to do them here.
```suggestion
		$widget_src = sprintf( 'https://widgets.wp.com/action-bar/#blog_id=%2$d&post_id=%3$d&origin=%1$s://%4$s', $protocol, $blog_id, $post_id, rawurlencode( $domain ) );
```

anomiex commented on projects/packages/action-bar/src/class-action-bar.php:
```diff
@@ -46,18 +47,111 @@ public function enqueue_scripts() {
+			$feed_id = \FeedBag::get_feed_id_for_blog_id( $blog_id );
+		}
+
+		if ( $feed_id ) {
+			return 'https://wordpress.com/read/feeds/' . esc_attr( $feed_id );
+		} else {
+			return 'https://wordpress.com/read/blogs/' . esc_attr( $blog_id );
+		}
```
Why the `esc_attr()` here? It should be handled by `esc_url()` when the URL is output. Here, where we're building the URL, I'd expect to see `rawurlencode()` instead.

anomiex commented on projects/packages/action-bar/src/js/action-bar.js:
```diff
@@ -1,29 +1,233 @@
+	// Manage focus
+	_on( document.querySelector( 'body' ), 'keydown', e => {
+		// Manually keep focus inside modal when tabbing
+		if ( 9 === e.which ) {
```
`e.which` is deprecated. Probably better to use `e.key` or `e.code` instead.

https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/which

anomiex reviewed (commented):
When testing, behavior seems inconsistent.

Most of the time, clicking the like or follow buttons opens a popup window, which quickly closes itself. Most of the time, nothing else seems to happen.

Occasionally I see this in my browser's console:
```
Failed to execute ‘postMessage’ on ‘DOMWindow’: The target origin provided (‘https://public-api.wordpress.com’) does not match the recipient window’s origin (‘https://widgets.wp.com’).
```
Oddly, after that happens then clicks seem to start doing something. Clicking the "like" star toggles it between blue and grey. Clicking the "follow" button tells me "Error following site", in which case the console seems more informative:
```
Follow site failed with code "already_subscribed", and message "You are already following that feed."
```
:shrug:

creativecoder commented:
Thanks for the feedback so far! I'm putting this PR on hold, for now, as the team is prioritizing another project for at least the next few weeks.

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into add/action-bar-iframed-like-button

anomiex commented:
Action Bar has been abandoned and removed from the monorepo. See #34939.
